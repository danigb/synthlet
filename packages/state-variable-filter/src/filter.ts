/**
 * Limit the value between min and max
 * @param value
 * @param min
 * @param max
 * @returns
 */
export function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

export enum SVFilterType {
  ByPass = 0,
  LowPass = 1,
  BandPass = 2,
  HighPass = 3,
}

export type Inputs = {
  filterType: number[];
  frequency: number[];
  resonance: number[];
};

/**
 * A State Variable Filter following the Andy Simper's implementation described in http://www.cytomic.com/files/dsp/SvfLinearTrapOptimised2.pdf
 *
 * Various implementations:
 * https://github.com/Ardour/ardour/blob/71e049202c017c0a546f39b455bdc9e4be182f06/libs/plugins/a-eq.lv2/a-eq.c
 * https://github.com/SoundStacks/cmajor/blob/main/standard_library/std_library_filters.cmajor#L157
 */
export function SVFilter(sampleRate: number) {
  // Params
  let $frequency = 1000;
  let $resonance = 0.5;
  let $type = SVFilterType.LowPass;

  const period = 0.5 / sampleRate;
  const pi2 = 2 * Math.PI;

  let d = 0;
  let a = 0;
  let g1 = 0;
  let z0 = 0;
  let z1 = 0;
  let high = 0;
  let band = 0;
  let low = 0;

  function update(inputs: Inputs) {
    $type = inputs.filterType[0];
    if (
      inputs.frequency[0] !== $frequency ||
      inputs.resonance[0] !== $resonance
    ) {
      $frequency = inputs.frequency[0];
      $resonance = inputs.resonance[0];
      const cutoffFreq = clamp($frequency, 16, sampleRate / 2);
      const Q = clamp($resonance, 0.025, 40);
      const invQ = 1.0 / Q;
      // TODO: review - something weird here (sampleRate * period = 0.5)
      a = 2.0 * sampleRate * Math.tan(pi2 * period * cutoffFreq) * period;
      d = 1.0 / (1.0 + invQ * a + a * a);
      g1 = a + invQ;
    }
  }

  function fill(input: Float32Array, output: Float32Array) {
    for (let i = 0; i < input.length; i++) {
      const x = input[i];
      high = (x - g1 * z0 - z1) * d;
      band = a * high + z0;
      low = a * band + z1;
      z0 = a * high + band;
      z1 = a * band + low;
      output[i] =
        $type === SVFilterType.LowPass
          ? low
          : $type === SVFilterType.HighPass
          ? high
          : $type === SVFilterType.BandPass
          ? band
          : x;
    }
  }
  return { update, fill };
}
