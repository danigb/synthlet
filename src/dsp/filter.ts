import { fastAtan } from "./math";

export class Filter {
  private band: number = 0.0;
  private low: number = 0.0;

  public isSilent(): boolean {
    return this.low === 0.0;
  }

  public process(
    dt: number,
    input: number,
    cutoff: number,
    resonance: number
  ): number {
    let f: number = 2.0 * Math.PI * cutoff * dt;
    f = f > 1 ? 1 : f < 0.01 ? 0.01 : f;

    // Resonance rolloff
    const maxResonance = 1.0 - f * f * f * f * f;
    const normalizedResonance =
      resonance > maxResonance ? maxResonance : resonance;

    const high = input - (this.low + this.band * (1 - normalizedResonance));
    this.band += f * high;
    this.low += f * this.band;
    this.low = fastAtan(this.low * 0.1) * 10.0;

    return this.low;
  }
}
