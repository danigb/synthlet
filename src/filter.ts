export class Filter {
  constructor() {}

  process(time: number) {}
}
