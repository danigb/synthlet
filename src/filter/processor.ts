export const PROCESSOR = `(()=>{function d(s){return Object.keys(s).map(e=>{let{min:t,max:r,defaultValue:o}=s[e];return{name:e,minValue:t,maxValue:r,defaultValue:o,automationRate:"k-rate"}})}var a=Math.PI,h=2*a,k=a*a,_=4/a,T=-4/k;var n=5/12,c=Math.pow(10,-1/n),C=1/(1-c),O=1/(1+n*Math.log10(1+c)),N=1/(1+-Math.pow(10,-1/n)),M=n*Math.log10(1+c),D=1/(-n*Math.log10(c)+M);var i=class{constructor(e){this.sampleRate=e;this.halfSamplePeriod=1/(2*e),this.cutoffFrequency=0,this.state=0,this.coeff=0,this.output={LPF1:0,HPF1:0,APF1:0,ANM_LPF1:0},this.update(1e3)}update(e){if(e===this.cutoffFrequency)return;this.cutoffFrequency=e;let t=Math.tan(h*this.cutoffFrequency*this.halfSamplePeriod);this.coeff=t/(1+t)}process(e){let t=(e-this.state)*this.coeff;return this.output.LPF1=(e-this.state)*this.coeff+this.state,this.output.HPF1=e-this.output.LPF1,this.output.APF1=this.output.LPF1-this.output.HPF1,this.output.ANM_LPF1=this.output.LPF1+this.coeff*this.output.HPF1,this.state=t+this.output.LPF1,this.output}};var p={filterType:{min:0,max:14,defaultValue:1},frequency:{min:0,max:1e4,defaultValue:1e3},resonance:{min:0,max:1,defaultValue:.5}},f=class{constructor(e){this.sampleRate=e;this.filterType=0,this.filterVa1=new i(e),this.process=t=>t,this.setParams(1,p.frequency.defaultValue,p.resonance.defaultValue)}setParams(e,t,r){this.filterType!==e&&this.#e(e),this.#t(t,r)}#e(e){switch(this.filterType=e,e){case 1:this.process=t=>this.filterVa1.process(t).LPF1;break;case 2:this.process=t=>this.filterVa1.process(t).HPF1;break;case 3:this.process=t=>this.filterVa1.process(t).APF1;break;default:this.process=t=>t;break}}#t(e,t){switch(this.filterType){case 1:case 2:case 3:this.filterVa1.update(e)}}};var P=class extends AudioWorkletProcessor{constructor(){super();this.impulse=new f(sampleRate)}process(t,r,o){let l=o.filterType[0],y=o.frequency[0],F=o.resonance[0];this.impulse.setParams(l,y,F);let x=t[0][0],b=r[0][0];for(let u=0;u<b.length;u++)b[u]=this.impulse.process(x[u]);return!0}static get parameterDescriptors(){return P.PARAMS}},m=P;m.PARAMS=d(p);registerProcessor("FilterWorklet",m);})();`;
