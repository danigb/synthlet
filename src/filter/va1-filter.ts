import { TWO_PI } from "../shared/math";

/**
 * Virtual Analog first order filter
 *
 * Outputs low pass, high pass, all pass and low pass with analog FNG
 */
export class VA1Filter {
  private halfSamplePeriod: number;
  private cutoffFrequency: number;
  private state: number;
  output: { LPF1: number; HPF1: number; APF1: number; ANM_LPF1: number };
  private coeff: number;

  constructor(public readonly sampleRate: number) {
    this.halfSamplePeriod = 1.0 / (2.0 * sampleRate);
    this.cutoffFrequency = 0.0;
    this.state = 0.0;
    this.coeff = 0.0;
    this.output = {
      LPF1: 0.0,
      HPF1: 0.0,
      APF1: 0.0,
      ANM_LPF1: 0.0,
    };
    this.update(1000);
  }

  update(cutoff: number) {
    if (cutoff === this.cutoffFrequency) return;
    this.cutoffFrequency = cutoff;
    const g = Math.tan(TWO_PI * this.cutoffFrequency * this.halfSamplePeriod);
    this.coeff = g / (1.0 + g);
  }

  process(xn: number) {
    const vn = (xn - this.state) * this.coeff;

    this.output.LPF1 = (xn - this.state) * this.coeff + this.state;
    this.output.HPF1 = xn - this.output.LPF1;
    this.output.APF1 = this.output.LPF1 - this.output.HPF1;
    this.output.ANM_LPF1 = this.output.LPF1 + this.coeff * this.output.HPF1;
    this.state = vn + this.output.LPF1;
    return this.output;
  }
}
