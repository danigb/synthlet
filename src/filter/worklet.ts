import { toWorkletParams } from "../worklet-utils";
import { Filter, FilterParams } from "./filter";

/**
 * Single input-output filter with selectable algorithms
 */
export class FilterWorklet extends AudioWorkletProcessor {
  impulse: Filter;
  static PARAMS = toWorkletParams(FilterParams);

  constructor() {
    super();
    this.impulse = new Filter(sampleRate);
  }

  process(
    inputs: Float32Array[][],
    outputs: Float32Array[][],
    parameters: any
  ) {
    const filterType = parameters.filterType[0];
    const frequency = parameters.frequency[0];
    const resonance = parameters.resonance[0];
    this.impulse.setParams(filterType, frequency, resonance);
    const input = inputs[0][0];
    const output = outputs[0][0];
    for (let i = 0; i < output.length; i++) {
      output[i] = this.impulse.process(input[i]);
    }

    return true;
  }

  static get parameterDescriptors() {
    return FilterWorklet.PARAMS;
  }
}

registerProcessor("FilterWorklet", FilterWorklet);
