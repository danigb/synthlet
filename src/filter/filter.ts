import { ParamsDef } from "../worklet-utils";
import { VA1Filter } from "./va1-filter";

export enum VAFilterAlgorithm {
  BypassFilter = 0,
  VF_LP = 1,
  VF_HP = 2,
  VF_AP = 3,
  SVF_LP = 4,
  SVF_HP = 5,
  SVF_BP = 6,
  SVF_BS = 7,
  Korg35_LP = 8,
  Korg35_HP = 9,
  Moog_LP1 = 10,
  Moog_LP2 = 11,
  Moog_LP3 = 12,
  Moog_LP4 = 13,
  Diode_LP4 = 14,
}

export const FilterParams: ParamsDef = {
  filterType: { min: 0, max: 14, defaultValue: 1 },
  frequency: { min: 0, max: 10000, defaultValue: 1000 },
  resonance: { min: 0, max: 1, defaultValue: 0.5 },
} as const;

export class Filter {
  filterType: VAFilterAlgorithm;
  filterVa1: VA1Filter;
  process: (x: number) => number;

  constructor(public readonly sampleRate: number) {
    this.filterType = VAFilterAlgorithm.BypassFilter;
    this.filterVa1 = new VA1Filter(sampleRate);
    this.process = (x: number) => x;
    this.setParams(
      VAFilterAlgorithm.VF_LP,
      FilterParams.frequency.defaultValue,
      FilterParams.resonance.defaultValue
    );
  }

  setParams(
    filterType: VAFilterAlgorithm,
    frequency: number,
    resonance: number
  ) {
    if (this.filterType !== filterType) {
      this.#setType(filterType);
    }
    this.#update(frequency, resonance);
  }

  #setType(filterType: VAFilterAlgorithm) {
    this.filterType = filterType;
    switch (filterType) {
      case VAFilterAlgorithm.VF_LP:
        this.process = (x: number) => this.filterVa1.process(x).LPF1;
        break;
      case VAFilterAlgorithm.VF_HP:
        this.process = (x: number) => this.filterVa1.process(x).HPF1;
        break;
      case VAFilterAlgorithm.VF_AP:
        this.process = (x: number) => this.filterVa1.process(x).APF1;
        break;
      default:
        this.process = (x: number) => x;
        break;
    }
  }

  #update(frequency: number, resonance: number) {
    switch (this.filterType) {
      case VAFilterAlgorithm.VF_LP:
      case VAFilterAlgorithm.VF_HP:
      case VAFilterAlgorithm.VF_AP:
        this.filterVa1.update(frequency);
    }
  }
}
