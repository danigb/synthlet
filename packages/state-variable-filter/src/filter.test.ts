import { SVFilter, SVFilterType } from "./filter";

describe("SVFilter", () => {
  it("filters the signal", () => {
    const filter = SVFilter(20, SVFilterType.LowPass);
    const input = new Float32Array(20);
    for (let i = 0; i < input.length; i++) {
      input[i] = i % 2 ? 1 : -1;
    }
    expect(input).toMatchSnapshot();
    const output = new Float32Array(20);
    filter.fill(input, output, {
      frequency: [10],
      resonance: [0.5],
    });
    expect(output).toMatchSnapshot();
    filter.fill(input, output, {
      frequency: [5],
      resonance: [0.5],
    });
    expect(output).toMatchSnapshot();
  });
});
