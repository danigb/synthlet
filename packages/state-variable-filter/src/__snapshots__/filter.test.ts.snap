// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVFilter filters the signal 1`] = `
Float32Array [
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
  -1,
  1,
]
`;

exports[`SVFilter filters the signal 2`] = `
Float32Array [
  -7.058988571166992,
  -96.1963882446289,
  -933.3245239257812,
  -7950.83740234375,
  -63193.6796875,
  -481030.5,
  -3555095.5,
  -25716442,
  -183016032,
  -1285889280,
  -8941918208,
  -61654106112,
  -422081658880,
  -2872091344896,
  -19441708433408,
  -131008733118464,
  -879299460595712,
  -5880895244337152,
  -39209108632502270,
  -260680291251650560,
]
`;

exports[`SVFilter filters the signal 3`] = `
Float32Array [
  -1728730658831859700,
  -11437933590730506000,
  -75519349223376030000,
  -497665280641740050000,
  -3.2738075880989456e+21,
  -2.150135330517005e+22,
  -1.4100311966854957e+23,
  -9.233990443912205e+23,
  -6.03935217634656e+24,
  -3.9452060773814655e+25,
  -2.574315262255496e+26,
  -1.6780219466165328e+27,
  -1.0927115680685831e+28,
  -7.109042947525591e+28,
  -4.621019964675439e+29,
  -3.001284075478557e+30,
  -1.9477740115635436e+31,
  -1.263136567931854e+32,
  -8.18576013175832e+32,
  -5.3012721012231e+33,
]
`;
